game/kernel/asm_funcs_arm64.s

```diff
@@ -0,0 +1,262 @@
+  ldp x12, x13, [sp], #32
+  ldp x14, x15, [sp], #32
+
+  ret
+
+
+;; Call C++ code on arm64 systems, from GOAL. 
+;; 
+;; Put arguments on the stack and put a pointer to this array in the first arg.
+;; this function pushes all 8 OpenGOAL registers into a stack array.
+;; then it calls the function pointed to by x0 (RAX in x86) with a pointer to this array.
+;; it returns the return value of the called function.
+.global _stack_call_arm64
+.align 4
+_stack_call_arm64:
+  ldr x8, [sp], #16
+
+  stp q15, q14, [sp, #-32]!
+  stp q13, q12, [sp, #-32]!
+  stp q11, q10, [sp, #-32]!
+  stp q9, q8, [sp, #-32]!
+
+  ; create stack array of arguments
+  ; arg 7 (R11 in x86)
+  ; arg 6 (R10 in x86)
+  ; arg 5 (R8 in x86)
+  ; arg 4 (R8 in x86)
+  ; arg 3 (RCX in x86)
+  ; arg 2 (RDX in x86)
+  ; arg 1 (RSI in x86)
+  ; arg 0 (RDI in x86)
+  stp x7, x6, [sp, #-16]!
+  stp x5, x4, [sp, #-16]!
+  stp x3, x2, [sp, #-16]!
+  stp x1, x0, [sp, #-16]!
+
+  ; set first argument
+  mov x19, sp
+  ; call function
+  blr x8
+  ; restore arguments
+  ldp x0, x1, [sp], #16
+  ldp x2, x3, [sp], #16
+  ldp x4, x5, [sp], #16
+  ldp x6, x7, [sp], #16
+
+  ldp x8, x9, [sp], #32
+  ldp x10, x11, [sp], #32
+  ldp x12, x13, [sp], #32
+  ldp x14, x15, [sp], #32
+
+  ; return!
+  ret
+
+;; Call c++ code through mips2c.
+;; GOAL will call a dynamically generated trampoline.
+;; The trampoline will have pushed the exec function and stack offset onto the stack
+.global _mips2c_call_arm64
+.align 4
+_mips2c_call_arm64:
+  ;; TODO - this is really weird using half an XMM, this makes the arm assembly
+  ;; more difficult - this probably isn't required for arm?
+  ;; grab the address to call and put it in xmm0
+  ;; TODO - this stack pointer manipulation might be a problem for ARM64 which requires 16byte alignment
+  ;; sub sp, 8
+  ldr q0, [sp, #+16]
+  ;; grab the stack offset
+  ldr x0, [sp, #+8]
+
+  ;; first, save quadword registers
+  stp q15, q14, [sp, #-32]!
+  stp q13, q12, [sp, #-32]!
+  stp q11, q10, [sp, #-32]!
+  stp q9, q8, [sp, #-32]!
+
+  ; NOTE - in x86 the 2 special registers are saved (R10 and R11)
+  ; we don't need to do that in ARM64, there are plenty of registers to work with
+
+  ;; oof
+  sub sp, sp, 1280
+  str x0, [sp, #+64] ; arg 0 (RDI in x86) and 
+  str x1, [sp, #+80] ; arg 1 (RSI in x86)
+  str x2, [sp, #+96] ; arg 2 (RDX in x86) and arg 3 (RCX in x86)
+  str x3, [sp, #+112] ; arg 2 (RDX in x86) and arg 3 (RCX in x86)
+  str x4, [sp, #+128] ; arg 4 (R8 in x86) and arg 5 (R8 in x86)
+  str x5, [sp, #+144] ; arg 4 (R8 in x86) and arg 5 (R8 in x86)
+  str x6, [sp, #+160] ; arg 6 (R10 in x86) and arg 7 (R11 in x86)
+  str x7, [sp, #+176] ; arg 6 (R10 in x86) and arg 7 (R11 in x86)
+  str x20, [sp, #+352] ;; s6 (pp) (R13 in x86) and s7 (st) (R14 in x86)
+  str x21, [sp, #+368] ;; s6 (pp) (R13 in x86) and s7 (st) (R14 in x86)
+
+  mov x0, sp ; move the stack pointer to arg 0
+  sub x0, x0, x22 ; R15 is a "special" offset TODO - whats special about it?
```

Comment:
This is the offset between real pointers, and GOAL pointers. The tricks are:
- when doing `call` or `blr` or load or store, the register needs to be a real pointer.
- Real pointers will be 64-bits
- GOAL pointers can only be 32-bits (they are stored in types as 32 bits only)
- To convert: `goal_pointer + offset = real_pointer` where `offset` is a 64-bit value.
- the stack pointer is a real pointer. This is just convenient so we can push/pop.
- in mips2c code, we need to set the emulated `sp` register's value to a goal pointer (to simulate what it would have been on ps2). So we subtract the offset here. The store below this is what stores `sp` into the simulated MIPS2C register struct.

Reply:
https://github.com/open-goal/jak-project/blob/master/game/mips2c/mips2c_private.h#L225